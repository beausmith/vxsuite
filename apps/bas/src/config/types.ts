// Events
export type EventTargetFunction = (event: React.FormEvent<EventTarget>) => void;
export type ButtonEvent = React.MouseEvent<HTMLButtonElement>;
export type ButtonEventFunction = (event: ButtonEvent) => void;
