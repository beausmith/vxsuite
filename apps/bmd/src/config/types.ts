import React from 'react'

// Generic
export interface Dictionary<T> {
  [key: string]: T | undefined
}

// Events
export type InputEvent = React.FormEvent<EventTarget>
export type ButtonEvent = React.MouseEvent<HTMLButtonElement>

// Election
export interface Candidate {
  readonly id: string
  readonly name: string
  readonly party?: string
  isWriteIn?: boolean
}
export type OptionalCandidate = Candidate | undefined
export type Vote = Candidate[] // will eventually add "| Approval" to this type
export type OptionalVote = Candidate[] | undefined
export type VotesDict = Dictionary<Vote>
export interface Contest {
  readonly id: string
  readonly section?: string
  readonly title: string
  readonly type: string
}
export interface CandidateContest extends Contest {
  readonly type: 'candidate'
  readonly seats: number
  readonly candidates: Candidate[]
  readonly allowWriteIns: boolean
}
export interface YesNoContest extends Contest {
  readonly type: 'yesno'
  readonly description: string
}
export interface BMDConfig {
  readonly requireActivation?: boolean
  readonly showHelpPage?: boolean
  readonly showSettingsPage?: boolean
}
export interface ElectionDefaults {
  readonly bmdConfig: BMDConfig
}
export interface Election {
  readonly contests: Array<CandidateContest | YesNoContest>
  readonly county: string
  readonly date: string
  readonly seal: string
  readonly state: string
  readonly title: string
  readonly bmdConfig?: BMDConfig
}
export type OptionalElection = Election | undefined
export type UpdateVoteFunction = (contestId: string, vote: Vote) => void
export interface BallotContextInterface {
  readonly election: Election | undefined
  resetBallot: () => void
  setBallotKey: (activationCode: string) => void
  updateVote: UpdateVoteFunction
  votes: VotesDict
}

export default {}
