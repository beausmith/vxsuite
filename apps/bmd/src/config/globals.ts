export default {
  FONT_SIZES: [18, 24, 28, 32],
  TEXT_SIZE: 1,
}
