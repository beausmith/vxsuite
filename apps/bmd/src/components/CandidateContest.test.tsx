// import { axe } from 'jest-axe'
import React from 'react'
import { fireEvent, render } from 'react-testing-library'

import { CandidateContest as CandidateContestInterface } from '../config/types'

import electionSample from '../data/electionSample.json'

import CandidateContest from './CandidateContest'

const contest = {
  allowWriteIns: true,
  candidates: [
    {
      id: 'solis',
      name: 'Andrea Solis',
      party: 'Federalist',
    },
    {
      id: 'keller',
      name: 'Amos Keller',
      party: "People's",
    },
    {
      id: 'rangel',
      name: 'Davitra Rangel',
      party: 'Liberty',
    },
  ],
  id: 'state-assembly',
  seats: 1,
  section: '54th District',
  title: 'State Assembly',
  type: 'candidate',
} as CandidateContestInterface
const candidate0 = contest.candidates[0]
const candidate1 = contest.candidates[1]
const candidate2 = contest.candidates[2]

it(`allows any candidate to be selected when no candidate is selected`, () => {
  const updateVote = jest.fn()
  const { container, getByText } = render(
    <CandidateContest contest={contest} vote={[]} updateVote={updateVote} />
  )
  expect(container).toMatchSnapshot()

  fireEvent.click(getByText(candidate0.name).closest('label')!)
  expect(updateVote).toHaveBeenCalledTimes(1)

  fireEvent.click(getByText(candidate1.name).closest('label')!)
  expect(updateVote).toHaveBeenCalledTimes(2)

  fireEvent.click(getByText(candidate2.name).closest('label')!)
  expect(updateVote).toHaveBeenCalledTimes(3)
})

it(`doesn't allow other candidates to be selected when 1 of 1 candidates have been selected`, () => {
  const updateVote = jest.fn()
  const { container, getByText, queryByText } = render(
    <CandidateContest
      contest={contest}
      vote={[candidate0]}
      updateVote={updateVote}
    />
  )
  expect(container).toMatchSnapshot()

  const candidate0Input = getByText(candidate0.name)
    .closest('label')!
    .querySelector('input')!
  expect(candidate0Input.disabled).toBeFalsy()
  expect(candidate0Input.checked).toBeTruthy()

  const candidate1Input = getByText(candidate1.name)
    .closest('label')!
    .querySelector('input')!
  expect(candidate1Input.disabled).toBeTruthy()
  expect(candidate1Input.checked).toBeFalsy()

  const candidate2Input = getByText(candidate2.name)
    .closest('label')!
    .querySelector('input')!
  expect(candidate2Input.disabled).toBeTruthy()
  expect(candidate2Input.checked).toBeFalsy()

  expect(queryByText('add write-in candidate')).toBeFalsy()

  fireEvent.click(getByText(candidate1.name).closest('label')!)
  expect(updateVote).not.toHaveBeenCalled()

  fireEvent.click(getByText(candidate2.name).closest('label')!)
  expect(updateVote).not.toHaveBeenCalled()

  fireEvent.click(getByText(candidate0.name).closest('label')!)
  expect(updateVote).toHaveBeenCalled()
})

it(`allows another candidate to be selected when 1 of 2 have been selected`, () => {
  const updateVote = jest.fn()
  const { container, getByText, queryByText } = render(
    <CandidateContest
      contest={{ ...contest, seats: 2 }}
      vote={[candidate0]}
      updateVote={updateVote}
    />
  )
  expect(container).toMatchSnapshot()

  const candidate0Input = getByText(candidate0.name)
    .closest('label')!
    .querySelector('input')!
  expect(candidate0Input.disabled).toBeFalsy()
  expect(candidate0Input.checked).toBeTruthy()

  const candidate1Input = getByText(candidate1.name)
    .closest('label')!
    .querySelector('input')!
  expect(candidate1Input.disabled).toBeFalsy()
  expect(candidate1Input.checked).toBeFalsy()

  const candidate2Input = getByText(candidate2.name)
    .closest('label')!
    .querySelector('input')!
  expect(candidate2Input.disabled).toBeFalsy()
  expect(candidate2Input.checked).toBeFalsy()

  expect(queryByText('add write-in candidate')).toBeTruthy()

  fireEvent.click(getByText(candidate1.name).closest('label')!)
  expect(updateVote).toHaveBeenCalled()

  fireEvent.click(getByText(candidate2.name).closest('label')!)
  expect(updateVote).toHaveBeenCalled()

  fireEvent.click(getByText(candidate0.name).closest('label')!)
  expect(updateVote).toHaveBeenCalled()
})

const contestWithWriteIns = electionSample.contests.find(
  c => !!c.allowWriteIns && c.type === 'candidate'
) as CandidateContestInterface
it(`displays warning if write-in candidate name is too long`, () => {
  const updateVote = jest.fn()
  const { container, getByText } = render(
    <CandidateContest
      contest={contestWithWriteIns}
      vote={[]}
      updateVote={updateVote}
    />
  )
  fireEvent.click(getByText('add write-in candidate').closest('button')!)
  expect(getByText('Write-In Candidate')).toBeTruthy()
  Array.from('JACOB JOHANSON JINGLEHEIMMER SCHMIDTT').forEach(i => {
    const key = i === ' ' ? 'space' : i
    fireEvent.click(getByText(key).closest('button')!)
  })
  expect(
    getByText('You have entered 37 of maximum 40 characters.')
  ).toBeTruthy()
  expect(container).toMatchSnapshot()
})
