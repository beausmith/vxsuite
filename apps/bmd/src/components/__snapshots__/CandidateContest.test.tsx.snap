// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`supports multi-seat contests allows a second candidate to be selected when one is selected 1`] = `
.c2 {
  line-height: 1.2;
  max-width: 66ch;
}

.c2 h1 {
  line-height: 1.1;
  margin: 2rem 0 1rem;
}

.c2 h2 {
  margin: 1.5rem 0 1rem;
}

.c2 h3,
.c2 p {
  margin: 1rem 0;
}

.c2 h1 + h2 {
  margin-top: -0.75rem;
}

.c2 h1 + p,
.c2 h2 + p,
.c2 h3 + p {
  margin-top: -0.75rem;
}

.c2 :first-child {
  margin-top: 0;
}

.c2 :last-child {
  margin-bottom: 0;
}

.c3 {
  font-size: 0.85rem;
  font-weight: 600;
  text-transform: uppercase;
}

.c0 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c1 {
  width: 100%;
  max-width: 35rem;
  margin: 0px auto;
  padding: 1rem 0.75rem 0.5rem;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  position: relative;
  overflow: auto;
}

.c5 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  overflow: auto;
}

.c5:before {
  content: '';
  z-index: 1;
  -webkit-transition: opacity 0.25s ease;
  transition: opacity 0.25s ease;
  height: 5px;
  width: 100%;
  position: fixed;
  opacity: 0;
  background: linear-gradient( to bottom, rgb(177,186,190) 0%, transparent 100% );
}

.c6 {
  display: grid;
  grid-auto-rows: minmax(auto,1fr);
  grid-gap: 0.75rem;
  width: 100%;
  max-width: 35rem;
  margin: 0 auto;
  padding: 0.5rem 0.5rem 2rem;
}

.c9 {
  cursor: pointer;
  position: relative;
  display: grid;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 0.125rem;
  background: white;
  box-shadow: 0 0.125rem 0.125rem 0 rgba(0,0,0,0.14), 0 0.1875rem 0.0625rem -0.125rem rgba(0,0,0,0.12), 0 0.0625rem 0.3125rem 0 rgba(0,0,0,0.2);
  -webkit-transition: background 0.25s,color 0.25s;
  transition: background 0.25s,color 0.25s;
}

button.c9 {
  text-align: left;
}

.c9:focus-within {
  outline: -webkit-focus-ring-color auto 5px;
}

.c9:before {
  content: '';
  position: absolute;
  left: 0;
  top: 0;
  bottom: 0;
  background: white;
  border-right: 1px solid;
  border-color: lightgrey;
  width: 3rem;
  text-align: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-size: 2rem;
  border-radius: 0.125rem 0 0 0.125rem;
  color: #028099;
}

.c9 > div {
  word-break: break-word;
  padding: 0.5rem 0.5rem 0.5rem 4rem;
}

.c7 {
  cursor: pointer;
  position: relative;
  display: grid;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 0.125rem;
  background: #028099;
  color: white;
  box-shadow: 0 0.125rem 0.125rem 0 rgba(0,0,0,0.14), 0 0.1875rem 0.0625rem -0.125rem rgba(0,0,0,0.12), 0 0.0625rem 0.3125rem 0 rgba(0,0,0,0.2);
  -webkit-transition: background 0.25s,color 0.25s;
  transition: background 0.25s,color 0.25s;
}

button.c7 {
  text-align: left;
}

.c7:focus-within {
  outline: -webkit-focus-ring-color auto 5px;
}

.c7:before {
  content: 'âœ“';
  position: absolute;
  left: 0;
  top: 0;
  bottom: 0;
  background: white;
  border-right: 1px solid;
  border-color: #028099;
  width: 3rem;
  text-align: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-size: 2rem;
  border-radius: 0.125rem 0 0 0.125rem;
  color: #028099;
}

.c7 > div {
  word-break: break-word;
  padding: 0.5rem 0.5rem 0.5rem 4rem;
}

.c8 {
  margin-right: 0.5rem;
}

@media (min-width:480px) {
  .c2 {
    line-height: 1.3;
  }
}

@media (min-width:640px) {
  .c1 {
    padding: 1rem 1.5rem 0.5rem;
    padding-left: 5rem;
  }
}

@media (min-width:720px) {
  .c6 {
    padding-right: 1rem;
    padding-left: 1rem;
  }
}

@media (min-width:480px) {
  .c9 > div {
    padding: 1rem 1rem 1rem inherit;
  }
}

@media (min-width:480px) {
  .c7 > div {
    padding: 1rem 1rem 1rem inherit;
  }
}

<div>
  <main
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        <h1
          aria-label="City, Mayor."
        >
          <div
            class="c3"
          >
            City
          </div>
          Mayor
        </h1>
        <p>
          <strong>
            Vote for 
            2
            .
          </strong>
           You have selected
           
          1
          .
        </p>
      </div>
    </div>
    <div
      class="c4"
    >
      <div
        class="c5"
      >
        <div
          class="c6"
        >
          <label
            class="c7"
            for="name-0"
          >
            <input
              checked=""
              class="visually-hidden c8"
              id="name-0"
              name="contest-id"
              type="checkbox"
              value="name-0"
            />
            <div
              class="c2"
            >
              <p
                aria-label="Name 0, Party 0."
              >
                <strong>
                  Name 0
                </strong>
                <br />
                Party 0
              </p>
            </div>
          </label>
          <label
            class="c9"
            for="name-1"
          >
            <input
              class="visually-hidden c8"
              id="name-1"
              name="contest-id"
              type="checkbox"
              value="name-1"
            />
            <div
              class="c2"
            >
              <p
                aria-label="Name 1, Party 1."
              >
                <strong>
                  Name 1
                </strong>
                <br />
                Party 1
              </p>
            </div>
          </label>
          <label
            class="c9"
            for="name-2"
          >
            <input
              class="visually-hidden c8"
              id="name-2"
              name="contest-id"
              type="checkbox"
              value="name-2"
            />
            <div
              class="c2"
            >
              <p
                aria-label="Name 2, Party 0."
              >
                <strong>
                  Name 2
                </strong>
                <br />
                Party 0
              </p>
            </div>
          </label>
        </div>
      </div>
    </div>
  </main>
</div>
`;

exports[`supports single-seat contest allows any candidate to be selected when no candidate is selected 1`] = `
.c2 {
  line-height: 1.2;
  max-width: 66ch;
}

.c2 h1 {
  line-height: 1.1;
  margin: 2rem 0 1rem;
}

.c2 h2 {
  margin: 1.5rem 0 1rem;
}

.c2 h3,
.c2 p {
  margin: 1rem 0;
}

.c2 h1 + h2 {
  margin-top: -0.75rem;
}

.c2 h1 + p,
.c2 h2 + p,
.c2 h3 + p {
  margin-top: -0.75rem;
}

.c2 :first-child {
  margin-top: 0;
}

.c2 :last-child {
  margin-bottom: 0;
}

.c3 {
  font-size: 0.85rem;
  font-weight: 600;
  text-transform: uppercase;
}

.c0 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c1 {
  width: 100%;
  max-width: 35rem;
  margin: 0px auto;
  padding: 1rem 0.75rem 0.5rem;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  position: relative;
  overflow: auto;
}

.c5 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  overflow: auto;
}

.c5:before {
  content: '';
  z-index: 1;
  -webkit-transition: opacity 0.25s ease;
  transition: opacity 0.25s ease;
  height: 5px;
  width: 100%;
  position: fixed;
  opacity: 0;
  background: linear-gradient( to bottom, rgb(177,186,190) 0%, transparent 100% );
}

.c6 {
  display: grid;
  grid-auto-rows: minmax(auto,1fr);
  grid-gap: 0.75rem;
  width: 100%;
  max-width: 35rem;
  margin: 0 auto;
  padding: 0.5rem 0.5rem 2rem;
}

.c7 {
  cursor: pointer;
  position: relative;
  display: grid;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 0.125rem;
  background: white;
  box-shadow: 0 0.125rem 0.125rem 0 rgba(0,0,0,0.14), 0 0.1875rem 0.0625rem -0.125rem rgba(0,0,0,0.12), 0 0.0625rem 0.3125rem 0 rgba(0,0,0,0.2);
  -webkit-transition: background 0.25s,color 0.25s;
  transition: background 0.25s,color 0.25s;
}

button.c7 {
  text-align: left;
}

.c7:focus-within {
  outline: -webkit-focus-ring-color auto 5px;
}

.c7:before {
  content: '';
  position: absolute;
  left: 0;
  top: 0;
  bottom: 0;
  background: white;
  border-right: 1px solid;
  border-color: lightgrey;
  width: 3rem;
  text-align: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-size: 2rem;
  border-radius: 0.125rem 0 0 0.125rem;
  color: #028099;
}

.c7 > div {
  word-break: break-word;
  padding: 0.5rem 0.5rem 0.5rem 4rem;
}

.c8 {
  margin-right: 0.5rem;
}

@media (min-width:480px) {
  .c2 {
    line-height: 1.3;
  }
}

@media (min-width:640px) {
  .c1 {
    padding: 1rem 1.5rem 0.5rem;
    padding-left: 5rem;
  }
}

@media (min-width:720px) {
  .c6 {
    padding-right: 1rem;
    padding-left: 1rem;
  }
}

@media (min-width:480px) {
  .c7 > div {
    padding: 1rem 1rem 1rem inherit;
  }
}

<div>
  <main
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        <h1
          aria-label="City, Mayor."
        >
          <div
            class="c3"
          >
            City
          </div>
          Mayor
        </h1>
        <p>
          <strong>
            Vote for 
            1
            .
          </strong>
           You have selected
           
          0
          .
        </p>
      </div>
    </div>
    <div
      class="c4"
    >
      <div
        class="c5"
      >
        <div
          class="c6"
        >
          <label
            class="c7"
            for="name-0"
          >
            <input
              class="visually-hidden c8"
              id="name-0"
              name="contest-id"
              type="checkbox"
              value="name-0"
            />
            <div
              class="c2"
            >
              <p
                aria-label="Name 0, Party 0."
              >
                <strong>
                  Name 0
                </strong>
                <br />
                Party 0
              </p>
            </div>
          </label>
          <label
            class="c7"
            for="name-1"
          >
            <input
              class="visually-hidden c8"
              id="name-1"
              name="contest-id"
              type="checkbox"
              value="name-1"
            />
            <div
              class="c2"
            >
              <p
                aria-label="Name 1, Party 1."
              >
                <strong>
                  Name 1
                </strong>
                <br />
                Party 1
              </p>
            </div>
          </label>
          <label
            class="c7"
            for="name-2"
          >
            <input
              class="visually-hidden c8"
              id="name-2"
              name="contest-id"
              type="checkbox"
              value="name-2"
            />
            <div
              class="c2"
            >
              <p
                aria-label="Name 2, Party 0."
              >
                <strong>
                  Name 2
                </strong>
                <br />
                Party 0
              </p>
            </div>
          </label>
        </div>
      </div>
    </div>
  </main>
</div>
`;

exports[`supports single-seat contest doesn't allow other candidates to be selected when a candidate is selected 1`] = `
.c2 {
  line-height: 1.2;
  max-width: 66ch;
}

.c2 h1 {
  line-height: 1.1;
  margin: 2rem 0 1rem;
}

.c2 h2 {
  margin: 1.5rem 0 1rem;
}

.c2 h3,
.c2 p {
  margin: 1rem 0;
}

.c2 h1 + h2 {
  margin-top: -0.75rem;
}

.c2 h1 + p,
.c2 h2 + p,
.c2 h3 + p {
  margin-top: -0.75rem;
}

.c2 :first-child {
  margin-top: 0;
}

.c2 :last-child {
  margin-bottom: 0;
}

.c3 {
  font-size: 0.85rem;
  font-weight: 600;
  text-transform: uppercase;
}

.c0 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c1 {
  width: 100%;
  max-width: 35rem;
  margin: 0px auto;
  padding: 1rem 0.75rem 0.5rem;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  position: relative;
  overflow: auto;
}

.c5 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  overflow: auto;
}

.c5:before {
  content: '';
  z-index: 1;
  -webkit-transition: opacity 0.25s ease;
  transition: opacity 0.25s ease;
  height: 5px;
  width: 100%;
  position: fixed;
  opacity: 0;
  background: linear-gradient( to bottom, rgb(177,186,190) 0%, transparent 100% );
}

.c6 {
  display: grid;
  grid-auto-rows: minmax(auto,1fr);
  grid-gap: 0.75rem;
  width: 100%;
  max-width: 35rem;
  margin: 0 auto;
  padding: 0.5rem 0.5rem 2rem;
}

.c9 {
  cursor: pointer;
  position: relative;
  display: grid;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 0.125rem;
  background: white;
  box-shadow: 0 0.125rem 0.125rem 0 rgba(0,0,0,0.14), 0 0.1875rem 0.0625rem -0.125rem rgba(0,0,0,0.12), 0 0.0625rem 0.3125rem 0 rgba(0,0,0,0.2);
  -webkit-transition: background 0.25s,color 0.25s;
  transition: background 0.25s,color 0.25s;
}

button.c9 {
  text-align: left;
}

.c9:focus-within {
  outline: -webkit-focus-ring-color auto 5px;
}

.c9:before {
  content: '';
  position: absolute;
  left: 0;
  top: 0;
  bottom: 0;
  background: white;
  border-right: 1px solid;
  border-color: lightgrey;
  width: 3rem;
  text-align: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-size: 2rem;
  border-radius: 0.125rem 0 0 0.125rem;
  color: #028099;
}

.c9 > div {
  word-break: break-word;
  padding: 0.5rem 0.5rem 0.5rem 4rem;
}

.c7 {
  cursor: pointer;
  position: relative;
  display: grid;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 0.125rem;
  background: #028099;
  color: white;
  box-shadow: 0 0.125rem 0.125rem 0 rgba(0,0,0,0.14), 0 0.1875rem 0.0625rem -0.125rem rgba(0,0,0,0.12), 0 0.0625rem 0.3125rem 0 rgba(0,0,0,0.2);
  -webkit-transition: background 0.25s,color 0.25s;
  transition: background 0.25s,color 0.25s;
}

button.c7 {
  text-align: left;
}

.c7:focus-within {
  outline: -webkit-focus-ring-color auto 5px;
}

.c7:before {
  content: 'âœ“';
  position: absolute;
  left: 0;
  top: 0;
  bottom: 0;
  background: white;
  border-right: 1px solid;
  border-color: #028099;
  width: 3rem;
  text-align: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-size: 2rem;
  border-radius: 0.125rem 0 0 0.125rem;
  color: #028099;
}

.c7 > div {
  word-break: break-word;
  padding: 0.5rem 0.5rem 0.5rem 4rem;
}

.c8 {
  margin-right: 0.5rem;
}

@media (min-width:480px) {
  .c2 {
    line-height: 1.3;
  }
}

@media (min-width:640px) {
  .c1 {
    padding: 1rem 1.5rem 0.5rem;
    padding-left: 5rem;
  }
}

@media (min-width:720px) {
  .c6 {
    padding-right: 1rem;
    padding-left: 1rem;
  }
}

@media (min-width:480px) {
  .c9 > div {
    padding: 1rem 1rem 1rem inherit;
  }
}

@media (min-width:480px) {
  .c7 > div {
    padding: 1rem 1rem 1rem inherit;
  }
}

<div>
  <main
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        <h1
          aria-label="City, Mayor."
        >
          <div
            class="c3"
          >
            City
          </div>
          Mayor
        </h1>
        <p>
          <strong>
            Vote for 
            1
            .
          </strong>
           You have selected
           
          1
          .
        </p>
      </div>
    </div>
    <div
      class="c4"
    >
      <div
        class="c5"
      >
        <div
          class="c6"
        >
          <label
            class="c7"
            for="name-0"
          >
            <input
              checked=""
              class="visually-hidden c8"
              id="name-0"
              name="contest-id"
              type="checkbox"
              value="name-0"
            />
            <div
              class="c2"
            >
              <p
                aria-label="Name 0, Party 0."
              >
                <strong>
                  Name 0
                </strong>
                <br />
                Party 0
              </p>
            </div>
          </label>
          <label
            class="c9"
            for="name-1"
          >
            <input
              class="visually-hidden c8"
              disabled=""
              id="name-1"
              name="contest-id"
              type="checkbox"
              value="name-1"
            />
            <div
              class="c2"
            >
              <p
                aria-label="Name 1, Party 1."
              >
                <strong>
                  Name 1
                </strong>
                <br />
                Party 1
              </p>
            </div>
          </label>
          <label
            class="c9"
            for="name-2"
          >
            <input
              class="visually-hidden c8"
              disabled=""
              id="name-2"
              name="contest-id"
              type="checkbox"
              value="name-2"
            />
            <div
              class="c2"
            >
              <p
                aria-label="Name 2, Party 0."
              >
                <strong>
                  Name 2
                </strong>
                <br />
                Party 0
              </p>
            </div>
          </label>
        </div>
      </div>
    </div>
  </main>
</div>
`;

exports[`supports write-in candidates displays warning if write-in candidate name is too long 1`] = `
.c2 {
  line-height: 1.2;
  max-width: 66ch;
}

.c2 h1 {
  line-height: 1.1;
  margin: 2rem 0 1rem;
}

.c2 h2 {
  margin: 1.5rem 0 1rem;
}

.c2 h3,
.c2 p {
  margin: 1rem 0;
}

.c2 h1 + h2 {
  margin-top: -0.75rem;
}

.c2 h1 + p,
.c2 h2 + p,
.c2 h3 + p {
  margin-top: -0.75rem;
}

.c2 :first-child {
  margin-top: 0;
}

.c2 :last-child {
  margin-bottom: 0;
}

.c3 {
  font-size: 0.85rem;
  font-weight: 600;
  text-transform: uppercase;
}

.c0 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c1 {
  width: 100%;
  max-width: 35rem;
  margin: 0px auto;
  padding: 1rem 0.75rem 0.5rem;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  position: relative;
  overflow: auto;
}

.c5 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  overflow: auto;
}

.c5:before {
  content: '';
  z-index: 1;
  -webkit-transition: opacity 0.25s ease;
  transition: opacity 0.25s ease;
  height: 5px;
  width: 100%;
  position: fixed;
  opacity: 0;
  background: linear-gradient( to bottom, rgb(177,186,190) 0%, transparent 100% );
}

.c6 {
  display: grid;
  grid-auto-rows: minmax(auto,1fr);
  grid-gap: 0.75rem;
  width: 100%;
  max-width: 35rem;
  margin: 0 auto;
  padding: 0.5rem 0.5rem 2rem;
}

.c7 {
  cursor: pointer;
  position: relative;
  display: grid;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 0.125rem;
  background: white;
  box-shadow: 0 0.125rem 0.125rem 0 rgba(0,0,0,0.14), 0 0.1875rem 0.0625rem -0.125rem rgba(0,0,0,0.12), 0 0.0625rem 0.3125rem 0 rgba(0,0,0,0.2);
  -webkit-transition: background 0.25s,color 0.25s;
  transition: background 0.25s,color 0.25s;
}

button.c7 {
  text-align: left;
}

.c7:focus-within {
  outline: -webkit-focus-ring-color auto 5px;
}

.c7:before {
  content: '';
  position: absolute;
  left: 0;
  top: 0;
  bottom: 0;
  background: white;
  border-right: 1px solid;
  border-color: lightgrey;
  width: 3rem;
  text-align: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-size: 2rem;
  border-radius: 0.125rem 0 0 0.125rem;
  color: #028099;
}

.c7 > div {
  word-break: break-word;
  padding: 0.5rem 0.5rem 0.5rem 4rem;
}

.c8 {
  margin-right: 0.5rem;
}

@media (min-width:480px) {
  .c2 {
    line-height: 1.3;
  }
}

@media (min-width:640px) {
  .c1 {
    padding: 1rem 1.5rem 0.5rem;
    padding-left: 5rem;
  }
}

@media (min-width:720px) {
  .c6 {
    padding-right: 1rem;
    padding-left: 1rem;
  }
}

@media (min-width:480px) {
  .c7 > div {
    padding: 1rem 1rem 1rem inherit;
  }
}

<div>
  <main
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        <h1
          aria-label="City, Mayor."
        >
          <div
            class="c3"
          >
            City
          </div>
          Mayor
        </h1>
        <p>
          <strong>
            Vote for 
            1
            .
          </strong>
           You have selected
           
          0
          .
        </p>
      </div>
    </div>
    <div
      class="c4"
    >
      <div
        class="c5"
      >
        <div
          class="c6"
        >
          <label
            class="c7"
            for="name-0"
          >
            <input
              class="visually-hidden c8"
              id="name-0"
              name="contest-id"
              type="checkbox"
              value="name-0"
            />
            <div
              class="c2"
            >
              <p
                aria-label="Name 0, Party 0."
              >
                <strong>
                  Name 0
                </strong>
                <br />
                Party 0
              </p>
            </div>
          </label>
          <label
            class="c7"
            for="name-1"
          >
            <input
              class="visually-hidden c8"
              id="name-1"
              name="contest-id"
              type="checkbox"
              value="name-1"
            />
            <div
              class="c2"
            >
              <p
                aria-label="Name 1, Party 1."
              >
                <strong>
                  Name 1
                </strong>
                <br />
                Party 1
              </p>
            </div>
          </label>
          <label
            class="c7"
            for="name-2"
          >
            <input
              class="visually-hidden c8"
              id="name-2"
              name="contest-id"
              type="checkbox"
              value="name-2"
            />
            <div
              class="c2"
            >
              <p
                aria-label="Name 2, Party 0."
              >
                <strong>
                  Name 2
                </strong>
                <br />
                Party 0
              </p>
            </div>
          </label>
          <button
            class="c7"
          >
            <div
              class="c2"
            >
              <p
                aria-label="add write-in candidate."
              >
                <em>
                  add write-in candidate
                </em>
              </p>
            </div>
          </button>
        </div>
      </div>
    </div>
  </main>
</div>
`;
